### PR TITLE
Feat/thread conversation history through RAG chat

### PR DESCRIPTION
**Pull request by LeoDev0**
## Summary

The chat was stateless: each question was sent to the LLM in isolation, so follow-ups like *"tell me more about the second one"* failed because the model had no prior turns and the retrieval embedding was computed from the standalone question.

This PR threads conversation history end-to-end:

- **Frontend** sends prior turns alongside the current question (capped at 10 messages, empty placeholders filtered out via `buildHistory`).
- **Backend** forwards history into the LLM as a structured `[system, user, assistant, …, user]` message array, replacing the old single-`UserMessage` prompt.
- **Retrieval** is augmented by `rewriteQueryForRetrieval`, which concatenates the last 2 user turns with the current question so vector search has the semantic anchor needed to disambiguate follow-ups (zero extra LLM call; helper isolated so it can be swapped to LLM-based rewrite later).
- **"New chat"** button next to the response-mode toggle clears state for clean multi-turn testing.
- **Validation** on the wire: Gin binding tags reject unknown roles and empty content with HTTP 400.

### Key changes

- `backend/pkg/types/models.go` — new `Message` / `MessageRole`; `QueryRequest` gains optional `History []Message`.
- `backend/internal/services/rag_pipeline.go` — new `buildSystemPrompt`, restructured `chatCompletionParams(systemPrompt, history, question)`, plus `trimHistory` and `rewriteQueryForRetrieval`. `Query` and `QueryStream` accept history.
- `backend/internal/handlers/query.go`, `query_stream.go`, `query_mock.go` — interface widened to forward history.
- `frontend/src/lib/api/history.ts` (new), `query.ts`, `queryStream.ts`, `queryOnce.ts` — extended signatures and request body.
- `frontend/src/app/page.tsx` — snapshots `messages` before `setMessages` adds the new pair, so history excludes the in-flight turn.

### Tests added

- `TestTrimHistory`, `TestRewriteQueryForRetrieval`, `TestChatCompletionParams_MessageOrder`
- `TestQuery_HistoryThreadedToLLMAndRetrieval`, `TestQuery_HistoryBeyondCapIsTrimmed`
- `TestQueryStream_HistoryThreaded`
- Handler-level cases: forwards history to pipeline, rejects invalid role, rejects empty content

## Test plan

- [x] `go vet ./... && gofmt -l . && go test ./...` (backend) — clean
- [x] `npm run lint` and `tsc --noEmit` (frontend) — clean
- [x] `curl` smoke against running backend: invalid role → 400, empty content → 400, valid history → 200
- [ ] Manual browser multi-turn: upload a doc with enumerable items, ask "what sections does the manual have?" then "tell me about the second one" — expect the answer to refer to the second section
- [ ] Verify `history` array in DevTools Network tab on `/api/query/stream` POST, including correctness on a 12+ turn conversation (`history.length === 10`)
- [ ] Toggle response mode to "Single response" and repeat the multi-turn flow on `/api/query`
- [ ] Click **New chat** and confirm the next request sends `history: []`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: backend/internal/services/rag_pipeline_test.go [hunk 1]
```diff
@@ -799,12 +801,272 @@ func TestQuery_ContextBuiltFromMultipleSources(t *testing.T) {
...
+	}
+}
+
+func TestChatCompletionParams_MessageOrder(t *testing.T) {
+	tests := []struct {
+		name          string
+		history       []types.Message
+		question      string
+		expectedRoles []string
+		expectedTexts []string
```
Comment (LeoDev0) [inline backend/internal/services/rag_pipeline_test.go]: Parallel `expectedRoles` / `expectedTexts` slices carry the same risk that `backend/CLAUDE.md` calls out for table-driven tests ("define named types instead of repeating verbose anonymous struct literals"): a silent off-by-one between the two slices would still pass row-by-row asserts. A paired type keeps each expected message self-contained:

```go
type expectedMessage struct {
    role string
    text string
}
```

Then each row reads `[]expectedMessage{{role: "system", text: "system-prompt"}, ...}` and the loop walks one slice instead of indexing two. `TestQuery_HistoryThreadedToLLMAndRetrieval` (around line 948) has the same parallel-array shape and would benefit from the same change.

File: backend/internal/services/rag_pipeline_test.go [hunk 2]
```diff
@@ -799,12 +801,272 @@ func TestQuery_ContextBuiltFromMultipleSources(t *testing.T) {
...
+		question string
+		expected expected
+	}{
+		{
+			name:     "no history sends single user message after system, embeds question only",
+			history:  nil,
+			question: "what is Go",
+			expected: expected{
+				messageRoles: []string{"system", "user"},
+				messageTexts: []string{"", "what is Go"},
```
Comment (LeoDev0) [inline backend/internal/services/rag_pipeline_test.go]: `messageTexts[0] = ""` is never actually compared against anything — the loop's `case "system":` branch only does `assert.NotNil(t, m.OfSystem)`. The placeholder reads like a checked expectation when it isn't, so a regression in `buildSystemPrompt` would slip past this test.

Either drop the placeholder so the table reflects what's verified, or have the loop assert the system message contains the document context. The mocked search returns a chunk with `Content: "ctx"`, so something like:

```go
case "system":
    assert.NotNil(t, m.OfSystem)
    assert.Contains(t, m.OfSystem.Content.OfString.Value, "ctx")
```

would make this case carry weight.

File: frontend/src/lib/api/history.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import type { ChatMessage, ApiMessage } from '@/types';
+
+export const MAX_HISTORY_TURNS = 10;
+
+export function buildHistory(messages: ChatMessage[]): ApiMessage[] {
+  const cleaned = messages
+    .filter(m => m.content.trim().length > 0)
+    .map<ApiMessage>(m => ({ role: m.role, content: m.content }));
+  return cleaned.slice(-MAX_HISTORY_TURNS);
+}
```
Comment (LeoDev0) [inline frontend/src/lib/api/history.ts:12]: Filtering on empty content can leave an orphaned user turn whose assistant pair was dropped — e.g. if a previous turn streamed `done` with no tokens, the assistant message exists in state with empty content. After the filter you get `[user1, user2, assistant2]`, and the backend then sends `[system, user, user, assistant, user]` to the LLM. DeepSeek/OpenAI tolerate non-alternating roles, but the whole point of the `[system, history…, user]` shape we picked was canonical alternation.

A small change keeps pairs consistent — when an empty-content assistant is encountered, drop the user it was paired with:

```suggestion
export function buildHistory(messages: ChatMessage[]): ApiMessage[] {
  const cleaned: ApiMessage[] = [];
  for (const m of messages) {
    if (m.content.trim().length === 0) {
      if (m.role === 'assistant' && cleaned.at(-1)?.role === 'user') cleaned.pop();
      continue;
    }
    cleaned.push({ role: m.role, content: m.content });
  }
  return cleaned.slice(-MAX_HISTORY_TURNS);
}
```

**Review by LeoDev0 — COMMENTED** [review]
Drive-by review on the conversation-history changes. Three suggestions on the test-table shape, an unverified expectation, and one frontend edge case that can break role alternation.

File: backend/internal/services/rag_pipeline_test.go (hunk 1 above)
Comment (LeoDev0) [reply backend/internal/services/rag_pipeline_test.go]: Nice suggestion. Done

File: backend/internal/services/rag_pipeline_test.go (hunk 2 above)
Comment (LeoDev0) [reply backend/internal/services/rag_pipeline_test.go]: Nice catch. Fixed
